File: Java/GameOfLife.java


public class GameOfLife {

    public static final int WIDTH = 45;
    public static final int HEIGHT = 27;
    public static final double aliveChanceOnSpawn = 0.2;
    public static final double FPS = 8.0;

    public static void main(String[] args) {
        Cell[][] cells = new Cell[WIDTH][HEIGHT];
        for (int x = 0; x < WIDTH; x++) {
            for (int y = 0; y < HEIGHT; y++) cells[x][y] = new Cell(aliveChanceOnSpawn);
        }
        while (true) {
            double start = System.nanoTime();

            draw(cells);
            updateNeighbors(cells);

            double sleepTime = (1_000_000_000.0/FPS - (System.nanoTime() - start))/1_000_000.0;
            try { Thread.sleep((long)((sleepTime > 0) ? sleepTime : 0)); }
            catch(InterruptedException e) { Thread.currentThread().interrupt(); }
        }
    }

    static void draw(Cell[][] cells) {
        System.out.print("\033[H\033[2J"); // won't work with windows cmd
        System.out.flush();
        for (int i = 0; i < WIDTH + 2; i++) System.out.print("#");
        System.out.println("");
        for (int y = 0; y < HEIGHT; y++) {
            System.out.print("#");
            for (int x = 0; x < WIDTH; x++) cells[x][y].drawCell();
            System.out.println("#");
        }
        for (int i = 0; i < WIDTH + 2; i++) System.out.print("#");
    }

    static void updateNeighbors(Cell[][] cells) {
        for (int x = 0; x < WIDTH; x++) {
            for (int y = 0; y < HEIGHT; y++) {
                cells[x][y].clearNeighbors();
				int[][] offsets = {{1, 0}, {1, 1}, {1, -1}, {0, 1}, {0, -1}, {-1, 0}, {-1, 1}, {-1, -1}};
				for (int[] offset : offsets) {
					try { cells[x][y].addNeighbor(cells[x + offset[0]][y + offset[1]].getAlive()); }
					catch (ArrayIndexOutOfBoundsException ignored) {}
				}
            }
        }
        for (int x = 0; x < WIDTH; x++) {
            for (int y = 0; y < HEIGHT; y++) cells[x][y].sync();
        }
    }
}

File: C/gameOfLife.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <stdio.h>  

#define WIDTH 45
#define HEIGHT 27
#define aliveChanceOnSpawn 0.2
#define DELAY 50


struct Cell {
    int alive;
    int neighbors;
};

void draw(struct Cell cells[WIDTH][HEIGHT]) {
    system("CLS");
    for (int i = 0; i < WIDTH + 2; i++) printf("#");
    printf("\n");
    for (int y = 0; y < HEIGHT; y++) {
        printf("#");
        for (int x = 0; x < WIDTH; x++) printf(cells[x][y].alive ? "@" : " ");
        printf("#\n");
    }
    for (int i = 0; i < WIDTH + 2; i++) printf("#");
}

void updateNeighbors(struct Cell cells[WIDTH][HEIGHT]) {
    for (int x = 0; x < WIDTH; x++) {
        for (int y = 0; y < HEIGHT; y++) {
            cells[x][y].neighbors = 0;
            int offsets[8][2] = {{1, 0}, {1, 1}, {1, -1}, {0, 1}, {0, -1}, {-1, 0}, {-1, 1}, {-1, -1}};
            for (int i = 0; i < 8; i++) {
                if (x + offsets[i][0] >= 0 && x + offsets[i][0] < WIDTH && y + offsets[i][1] >= 0 && y + offsets[i][1] < HEIGHT && cells[x + offsets[i][0]][y + offsets[i][1]].alive) {
                    cells[x][y].neighbors++;
                }
            }
        }
    }
    for (int x = 0; x < WIDTH; x++) {
        for (int y = 0; y < HEIGHT; y++) {
            if (cells[x][y].neighbors == 3) cells[x][y].alive = 1;
            else if (cells[x][y].neighbors < 2 || cells[x][y].neighbors > 3) cells[x][y].alive = 0;
        }
    }
}


int main() {
    srand(time(NULL));
    struct Cell cells[WIDTH][HEIGHT];
    for (int x = 0; x < WIDTH; x++) {
        for (int y = 0; y < HEIGHT; y++) {
            int alive = ((double)rand()/(double)RAND_MAX < aliveChanceOnSpawn) ? 1 : 0; 
            struct Cell tempCell = {alive, 0};
            cells[x][y] = tempCell;
        }
    }
    while (1) {
        draw(cells);
        updateNeighbors(cells);
        _sleep(DELAY);
    }
}